transformer_engine/pytorch/graph.py

```diff
@@ -314,17 +398,18 @@ def hook_fn(module, inputs, outputs):  # pylint: disable=unused-argument
             else:
                 # Capture backward graph for model chunk c_id, microbatch bwd_idx[-c_id-1]
                 m_chunk = -c_id - 1
-                for l_no in list(reversed(range(num_layers))):
-                    per_callable_bwd_idx = (m_chunk * num_microbatches * num_layers) + (
-                        bwd_idx[m_chunk] * num_layers + l_no
+                for l_no in list(reversed(range(_num_layers_per_chunk[m_chunk]))):
+                    per_callable_bwd_idx = (_prefix_num_layers[m_chunk] * num_microbatches) + (
+                        bwd_idx[m_chunk] * _num_layers_per_chunk[m_chunk] + l_no
                     )
                     static_input_surface = per_callable_static_input_surfaces[per_callable_bwd_idx]
                     static_outputs = per_callable_static_outputs[per_callable_bwd_idx]
                     bwd_graph = bwd_graphs[per_callable_bwd_idx]
                     # For now, assumes all static_outputs require grad
-                    static_grad_outputs = tuple(
-                        torch.empty_like(o) if o.requires_grad else None for o in static_outputs
-                    )
+                    if not io_memory_reduction or static_grad_outputs is None:
+                        static_grad_outputs = tuple(
+                            torch.empty_like(o) if o.requires_grad else None for o in static_outputs
+                        )
```

Comment:
Future reference: grad output is only used within backward, so we can reuse the same static buffers every time.